onnxruntime/test/providers/cpu/tensor/unsqueeze_op_test.cc

```diff
@@ -46,42 +46,83 @@ TEST(TensorOpTest, Unsqueeze_3) {
-  //will fail schema validation.
-  OpTester test("Unsqueeze", 12);
-
-  test.AddAttribute("axes", std::vector<int64_t>{2, 1, 0, 2});
-  test.AddInput<float>("input", {2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
-  test.AddOutput<float>("output", {1, 1, 1, 2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
-  test.Run(OpTester::ExpectResult::kExpectFailure,
-           "[ShapeInferenceError] 'axes' attribute must not contain any duplicates",
-           {kTensorrtExecutionProvider});  //TensorRT failed
+  {
+    OpTester test("Unsqueeze", 12); // opset 1-12 has axes attribute
+
+    test.AddAttribute("axes", std::vector<int64_t>{2, 1, 0, 2});
+    test.AddInput<float>("input", {2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.AddOutput<float>("output", {1, 1, 1, 2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.Run(OpTester::ExpectResult::kExpectFailure,
+             "[ShapeInferenceError] 'axes' attribute must not contain any duplicates",
+             {kTensorrtExecutionProvider});  //TensorRT failed
+  }
+  {
+    OpTester test("Unsqueeze", 13); // use latest opset with axis input
+
+    test.AddInput<float>("input", {2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.AddInput<int64_t>("axes", {4}, std::vector<int64_t>{2, 1, 0, 2}, true); //set as initializer to enable shape inference
+    test.AddOutput<float>("output", {1, 1, 1, 2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.Run(OpTester::ExpectResult::kExpectFailure,
+             "[ShapeInferenceError] 'axes' attribute must not contain any duplicates",
+             {kTensorrtExecutionProvider});  //TensorRT failed
+  }
 }
 
 TEST(TensorOpTest, Unsqueeze_OutOfRange) {
-  // This test is valid for opset 12.
-  // setting opset to -1 makes the test infra pick the latest available opset, so ort will pull the
-  // schema for that opset and do verification against it. Since opset 13 is enabled this test
-  // will fail schema validation.
-  OpTester test("Unsqueeze", 12);
-
-  test.AddAttribute("axes", std::vector<int64_t>{4});
-  test.AddInput<float>("input", {2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
-  test.AddOutput<float>("output", {2, 1, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
-  test.Run(OpTester::ExpectResult::kExpectFailure,
-           "[ShapeInferenceError] values in 'axes' are beyond the bounds of the computed output shape");
+  {
+    OpTester test("Unsqueeze", 12); // opset 1-12 has axes attribute
+    test.AddAttribute("axes", std::vector<int64_t>{4});
+    test.AddInput<float>("input", {2, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.AddOutput<float>("output", {2, 1, 3, 4}, std::vector<float>(2 * 3 * 4, 1.0f));
+    test.Run(OpTester::ExpectResult::kExpectFailure,
+             "[ShapeInferenceError] values in 'axes' are beyond the bounds of the computed output shape");
+  }
+  {
+    OpTester test("Unsqueeze", 13);  // use latest opset with axis input
```

Comment:
Shape inference check is strict for latest opset and lenient for older opsets. Right now this will work but once we have new opsets it will be best to pick latest opset for checking shape inference failures.